### PR TITLE
feat: add search_forecasting dag

### PR DESCRIPTION
**Pull request by jaredsnyder**
## Description

This PR adds the search forecast dag. 

## Related Tickets & Documents
* https://mozilla-hub.atlassian.net/browse/RS-1248

### CODE REVIEW

File: README.md
Comment (mikaeld) [inline README.md:49]: Forked workflows are for contributors without write permission on the repo, which represents less than 1% of contributions to the repo.

File: README.md
Comment (mikewilli) [inline README.md:48]: Thanks for updating the README! I think we should clarify that this section applies to contributors without write access to push direct branches to the repo. For contributors with write access, a direct branch is perfectly acceptable.

File: README.md
Comment (mikewilli) [inline README.md:49]: Typo fix:
```suggestion
The best way to create a Pull Request is to fork this repo, make your changes in the forked repo, and submit a pull request from the forked repo into the central repository.  More info on how to do this can be found [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).
```

File: README.md
Comment (mikewilli) [inline README.md:51]: Perhaps something like this would be more clear?
```suggestion
PR name and description must follow [Conventional Commit style](https://www.conventionalcommits.org/en/v1.0.0/) for the commit message and body, respectively.
```

File: README.md
Comment (mikewilli) [inline README.md:52]: Might be helpful to add a note that a contributor with write access will need to approve and merge the PR.

File: dags/search_forecasting.py
Comment (mikewilli) [inline dags/search_forecasting.py:40]: Is this correct? Run once per month on the 7th at 7:07?

Comment (jaredsnyder) [reply dags/search_forecasting.py:40]: Yeah :D

File: README.md
Comment (jaredsnyder) [reply README.md:52]: That is helpful, adding.

File: README.md
Comment (jaredsnyder) [reply README.md]: Fixed

File: README.md
Comment (jaredsnyder) [reply README.md:48]: I left the title for the section but clarified the forked PR is only necessary for those without write access.  Do you think that is sufficient?

File: README.md
Comment (jaredsnyder) [reply README.md]: Fixed

File: dags/search_forecasting.py
Comment (m-d-bowerman) [reply dags/search_forecasting.py:40]: 7:07 might be too specific, but it should be running once a month on the 7th, yeah.

Comment (jaredsnyder) [reply dags/search_forecasting.py:40]: my reviewers have no love of beauty, no appreciation of the silly, smgdh

**Comment by jaredsnyder** [comment]
When I run Airflow locally, I get this error (full log attached)
[dag_id=search_forecasting_run_id=manual__2024-06-28T13_10_15.666217+00_00_task_id=search_forecasting_search_forecasting_search_count_attempt=1.log](https://github.com/user-attachments/files/16030724/dag_id.search_forecasting_run_id.manual__2024-06-28T13_10_15.666217%2B00_00_task_id.search_forecasting_search_forecasting_search_count_attempt.1.log)
:

```
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base] Writing results to `moz-fx-data-shared_prod.search_derived.search_funnel_forecasts_v1`.
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base] Traceback (most recent call last):
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 2484, in load_table_from_file
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     response = self._do_multipart_upload(
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 3051, in _do_multipart_upload
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     response = upload.transmit(
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/requests/upload.py", line 153, in transmit
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     return _request_helpers.wait_and_retry(
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/requests/_request_helpers.py", line 178, in wait_and_retry
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     raise error
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/requests/_request_helpers.py", line 155, in wait_and_retry
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     response = func()
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/requests/upload.py", line 149, in retriable_request
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     self._process_response(result)
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/_upload.py", line 114, in _process_response
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     _helpers.require_status_code(response, (http.client.OK,), self._get_status_code)
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/_helpers.py", line 108, in require_status_code
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     raise common.InvalidResponse(
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base] google.resumable_media.common.InvalidResponse: ('Request failed with status code', 500, 'Expected one of', <HTTPStatus.OK: 200>)
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base] During handling of the above exception, another exception occurred:
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base] Traceback (most recent call last):
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/app/./kpi_forecasting.py", line 33, in <module>
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     main()
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/app/./kpi_forecasting.py", line 26, in main
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     model.write_results(**config.write_results)
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/app/kpi_forecasting/models/funnel_forecast.py", line 682, in write_results
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     job = client.load_table_from_dataframe(
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 2732, in load_table_from_dataframe
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     return self.load_table_from_file(
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 2488, in load_table_from_file
[2024-06-28, 13:33:58 UTC] {pod_manager.py:466} INFO - [base]     raise exceptions.from_http_response(exc.response)
2024-06-28T13:33:58.443796317Z 
2024-06-28T13:33:58.443817277Z 
[2024-06-28, 13:33:59 UTC] {pod_manager.py:483} INFO - [base] google.api_core.exceptions.InternalServerError: 500 POST https://bigquery.googleapis.com/upload/bigquery/v2/projects/moz-fx-data-shared_prod/jobs?uploadType=multipart: unknown error
[2024-06-28, 13:33:59 UTC] {pod_manager.py:511} WARNING - Pod search-forecasting-search-forecasting-search-count-u3xm4wp0 log read interrupted but container base still running
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base] Traceback (most recent call last):
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 2484, in load_table_from_file
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     response = self._do_multipart_upload(
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 3051, in _do_multipart_upload
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     response = upload.transmit(
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/requests/upload.py", line 153, in transmit
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     return _request_helpers.wait_and_retry(
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/requests/_request_helpers.py", line 178, in wait_and_retry
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     raise error
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/requests/_request_helpers.py", line 155, in wait_and_retry
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     response = func()
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/requests/upload.py", line 149, in retriable_request
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     self._process_response(result)
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/_upload.py", line 114, in _process_response
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     _helpers.require_status_code(response, (http.client.OK,), self._get_status_code)
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/resumable_media/_helpers.py", line 108, in require_status_code
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     raise common.InvalidResponse(
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base] google.resumable_media.common.InvalidResponse: ('Request failed with status code', 500, 'Expected one of', <HTTPStatus.OK: 200>)
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base] During handling of the above exception, another exception occurred:
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base] Traceback (most recent call last):
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/app/./kpi_forecasting.py", line 33, in <module>
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     main()
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/app/./kpi_forecasting.py", line 26, in main
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     model.write_results(**config.write_results)
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/app/kpi_forecasting/models/funnel_forecast.py", line 682, in write_results
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     job = client.load_table_from_dataframe(
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 2732, in load_table_from_dataframe
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     return self.load_table_from_file(
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]   File "/usr/local/lib/python3.10/site-packages/google/cloud/bigquery/client.py", line 2488, in load_table_from_file
[2024-06-28, 13:34:00 UTC] {pod_manager.py:466} INFO - [base]     raise exceptions.from_http_response(exc.response)
[2024-06-28, 13:34:00 UTC] {pod_manager.py:483} INFO - [base] google.api_core.exceptions.InternalServerError: 500 POST https://bigquery.googleapis.com/upload/bigquery/v2/projects/moz-fx-data-shared_prod/jobs?uploadType=multipart: unknown error
[2024-06-28, 13:34:00 UTC] {pod_manager.py:616} INFO - Pod search-forecasting-search-forecasting-search-count-u3xm4wp0 has phase Running
[2024-06-28, 13:34:03 UTC] {pod.py:912} INFO - Skipping deleting pod: search-forecasting-search-forecasting-search-count-u3xm4wp0
```

If I'm interpreting this correctly, it all runs fine until it gets to the write step, at which point it is failing to write (which is expected because I don't have write permissions)

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:19]: Could you please also include `telemetry-alerts@mozilla.com` here?

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:54]: since you're already creating the DAG with a context manager this line is not needed.
```suggestion
```

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:55]:
```suggestion
```

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:56]:
```suggestion
```

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:57]:
```suggestion
```

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:58]:
```suggestion
```

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:21]: Are you expecting for Airflow to retrospectively generate runs for the dates since this date?

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:23]:
```suggestion
    "email_on_retry": False,
```

Generally, we do not email on retry. The only valid use case for keeping this enabled would be if you actually expect someone to respond to this try failing immediately otherwise it's just noise.

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py:21]: No, the first day we'll want is 7-7-24.  Should I set start_date to that date?

File: README.md
Comment (mikewilli) [reply README.md:48]: I think the changes are great. Might be good to include a note about contributors with write access -- added a suggestion inline below.

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py:40]: It is independent.  The reasoning is that it is doing a monthly aggregates for the previous month, and we're having it run 7 days after the start of the current month so everything should be ready for it then.

Comment (jaredsnyder) [reply dags/search_forecasting.py:40]: I will defer to @m-d-bowerman for more details about the 7 day lag if you are interested

File: README.md
Comment (mikewilli) [inline README.md:51]:
```suggestion
For contributors without write access to the repo, the best way to create a Pull Request is to fork this repo, make your changes in the forked repo, and submit a pull request from the forked repo into the central repository.  More info on how to do this can be found [here](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request-from-a-fork).  In order to merge a PR, someone with write access must approve it.

Most contributors to this repo are likely to have write access. For these contributors, using a fork is not necessary and creating a branch in this repo directly is the preferred workflow. The normal PR process still applies.

For all PRs (forked or not), PR name and description must follow [Conventional Commit style](https://www.conventionalcommits.org/en/v1.0.0/).
```

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py]: fixed

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py]: fixed in latest commit

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py]: fixed in latest commit

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py]: fixed in latest commit

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py]: fixed in latest commit

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py]: fixed in latest commit

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py]: Whoops, fixed in latest commit

File: dags/search_forecasting.py
Comment (kik-kik) [reply dags/search_forecasting.py:25]: set it to one day before that date.

File: dags/search_forecasting.py
Comment (kik-kik) [reply dags/search_forecasting.py:44]: all good, I was just trying to check if we should be waiting for a specific table to update before running this. Technically you could still do that and just set the task to look for a run of the upstream job 7 days in the past.

**Review by kik-kik — APPROVED** [review]
I looked at the DAG changes, seems fine to me.

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py:44]: Oh that would be nice, how would I go about checking the upstream job ran 7 days ago?

File: dags/search_forecasting.py
Comment (jaredsnyder) [reply dags/search_forecasting.py]: Fixed

File: dags/search_forecasting.py
Comment (kik-kik) [reply dags/search_forecasting.py:44]: You can try looking at this example:
https://github.com/mozilla/telemetry-airflow/blob/04566b23cc3004ff5e61ef401510cf86dca459f5/dags/adm_export.py#L80-L90

`execution_delta` being the parameter you'd probably need to adjust to be 7 days.

https://airflow.apache.org/docs/apache-airflow/stable/howto/operator/external_task_sensor.html#externaltasksensor

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:48]: This needs to be moved inside the DAG context manager.

Comment (jaredsnyder) [reply dags/search_forecasting.py]: whoops, pushed a fix just now

File: dags/search_forecasting.py
Comment (kik-kik) [inline dags/search_forecasting.py:46]: I think for the sensor to work correctly you'll also need to update the schedule interval to match that of the external task. So I believe this should be the interval: `30 4 7 * *`

**Review by m-d-bowerman — APPROVED** [review]
lgtm

**Pull request merged**
